mahidhar/segtree.cpp

```diff
@@ -0,0 +1,106 @@
+//This is a generic implementation of the segment tree without lazy propagation
+//This solves range sum query with updates
+//For other problems the implementation of the util class should be changed accordingly.
+
+#include <iostream>
+
+struct sTreeNode{
+	int x;
+};
+
+typedef struct sTreeNode snode;
+
+class util{
+public:
+	//This is the return value if the range is out of bounds
+	snode outValue;
+
+	util(){
+		outValue.x = 0;
+	}
+	snode getLeaf(int a){
+		snode ans;
+		ans.x = a;
+		return ans;
+	}
+	snode combine(const snode &a,const snode &b){
+		snode ans;
+		ans.x = a.x+b.x;
+		return ans;
+	}
+	snode getVal(const snode &a){
+		snode ans;
+		ans.x = a.x;
+		return ans;
+	}
+};
+
+//IMPORTANT:: Assumes array is 1 indexed.
+class segtree{
+public:
+	int SIZE;
+	snode *seg;
+	util u;
+	//constructor
+	segtree(int size):
+	SIZE(size)
+	{
+		seg = new snode[4*size+1];
+		for(int i = 0 ; i <= 4*size ; i++)
+			seg[i].x = 0;
+	}
+
+	//function to build the tree
+	void build(int arr[],  int L , int R , int node = 1)
```

Comment:
L, R and M are not descriptive variable names.

Reply:
Perhaps arr should be a member of the class segment tree. Also consider using a vector instead of array using .reserve() to fix the size of the vector.